### PR TITLE
Update TSLint to latest v5.1.0

### LINKED ISSUE #57: Custom Rule: `no-semicolon-after-interface`
As per @GoodForOneFare 
> @ismail-syed can you take a look at linting away "no semicolon after interface". If there's an existing rule, let's use it.

```
interface foo {
  bar: string
};
 ^ interfaces should not be terminated by semicolons
```

### PR DESCRIPTION
**Pull request by ismail-syed**
* Updated `typescript` to `v2.3.2` 
* Updated `tslint` to `v5.1.0`
* Updated `tslint-react` to `v3.0.0`

Fixes #57 with `tslint` v5.1.0 update

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,17 @@
 
 ## [Unreleased]
 
+## [2.0.0] - 2017-05-02
+### Added
+- Added new `tslint` [v5.0.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v500), [v5.1.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v510) rules and new `tslint-react`[v2.6.0](https://github.com/palantir/tslint-react/releases/tag/2.6.0) rules [(#61)](https://github.com/Shopify/tslint-config-shopify/pull/61)
```
Comment (lemonmade) [inline CHANGELOG.md]: Weird sentence, and missing a space after `tslint-react` I think

File: src/rules/maintainability.ts [hunk 2]
```diff
@@ -25,6 +25,8 @@ export default {
   'no-require-imports': true,
   // Disallows trailing whitespace at the end of a line.
   'no-trailing-whitespace': true,
+  // Replaces x => f(x) with just f. To catch more cases, enable only-arrow-functions and arrow-return-shorthand too.
+  'no-unnecessary-callback-wrapper': false,
```
Comment (lemonmade) [inline src/rules/maintainability.ts]: Why `false`?

File: src/rules/react.ts [hunk 3]
```diff
@@ -12,6 +12,8 @@ export default {
   'jsx-boolean-value': [true, 'never'],
   // Requires or bans spaces between curly brace characters in JSX; Rule options: ["always", "never"]
   'jsx-curly-spacing': ['never'],
+  // Forbids function binding in JSX attributes. This has the same intent as jsx-no-lambda in helping you avoid excessive re-rendres.
+  'jsx-no-bind': false,
```
Comment (lemonmade) [inline src/rules/react.ts]: Why `false` here?

File: src/rules/style.ts [hunk 4]
```diff
@@ -43,6 +43,8 @@ export default {
   'no-consecutive-blank-lines': [true, 2],
   // Disallows parameter properties in class constructors.
   'no-parameter-properties': false,
+  // Don’t if you import “foo” anyway.
```
Comment (lemonmade) [inline src/rules/style.ts:46]: this sentence doesn't seem to make sense? As a result I don't understand the rule

**Review by lemonmade — CHANGES_REQUESTED** [review]

File: src/rules/style.ts (hunk 4 above)
Comment (ismail-syed) [reply src/rules/style.ts:46]: No idea either  😐 
https://palantir.github.io/tslint/rules/no-reference-import/

File: src/rules/maintainability.ts (hunk 2 above)
Comment (ismail-syed) [reply src/rules/maintainability.ts]: Left this up for discussion, I'll turn this on.

File: src/rules/react.ts (hunk 3 above)
Comment (ismail-syed) [reply src/rules/react.ts]: Same here, I'll turn it on.

File: src/rules/style.ts (hunk 4 above)
Comment (lemonmade) [reply src/rules/style.ts:46]: lol wtf. My guess would be that it means to not use a `/// reference` comment for things that are already imported. I would just turn it on for now and we can disable if it is doing something unexpected.

File: src/rules/maintainability.ts (hunk 2 above)
Comment (ismail-syed) [reply src/rules/maintainability.ts]: Fixed: 03f5c03

File: src/rules/react.ts (hunk 3 above)
Comment (ismail-syed) [reply src/rules/react.ts]: Fixed 03f5c03

File: CHANGELOG.md (hunk 1 above)
Comment (ismail-syed) [reply CHANGELOG.md]: Fixed: 951c656

File: src/rules/style.ts (hunk 4 above)
Comment (ismail-syed) [reply src/rules/style.ts:46]: Yup, that what it seems to be doing. 
https://github.com/palantir/tslint/blob/master/test/rules/no-reference-import/test.d.ts.lint

Comment (ismail-syed) [reply src/rules/style.ts:46]: Fixed: da1c74e

**Comment by ismail-syed** [comment]
@lemonmade Fixes are up 👍

File: src/customRules/trailingCommaInterfaceRule.ts [hunk 5]
```diff
@@ -1,6 +1,19 @@
 import * as ts from 'typescript';
 import * as Lint from 'tslint';
 
+class TrailingCommaInterfaceWalker extends Lint.RuleWalker {
```
Comment (GoodForOneFare) [inline src/customRules/trailingCommaInterfaceRule.ts:4]: Convention in `tslint` is to put the walker after the rule (e.g., [unified-signatures-rule](https://github.com/palantir/tslint/blob/b4be183b8e3d9fa2e07d7e0ab34c500ee0b0bb7b/src/rules/unifiedSignaturesRule.ts#L59)).

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,19 @@
 
 ## [Unreleased]
 
+## [2.0.0] - 2017-05-02
+### Added
+- Added new `tslint` rules from [v5.0.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v500) and [v5.1.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v510) releases
```
Comment (GoodForOneFare) [inline CHANGELOG.md]: Just reading this, I'd assume that all of these new rules are enabled.

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,19 @@
...
+    * `no-invalid-template-strings`
+    * `no-sparse-arrays`
+    * `no-unnecessary-callback-wrapper`
+    * `no-reference-import`
+    * `prefer-template`
+    * `return-undefined`
+- Added new `tslint-react` rules from [v2.6.0](https://github.com/palantir/tslint-react/releases/tag/2.6.0) release [(#61)](https://github.com/Shopify/tslint-config-shopify/pull/61)
+    * `jsx-no-bind` (tslint-react)
+
+
```
Comment (GoodForOneFare) [inline CHANGELOG.md]: Should mention `no-unused-new`'s removal

File: src/rules/functionality.ts [hunk 8]
```diff
@@ -53,6 +55,8 @@ export default {
   'no-null-keyword': false,
   // Disallows shadowing variable declarations.
   'no-shadowed-variable': true,
+  // Forbids array literals to contain missing elements.
+  'no-sparse-arrays': false,
```
Comment (GoodForOneFare) [inline src/rules/functionality.ts]: No sparse arrays in Neutron.  What's the value of allowing them?

**Review by GoodForOneFare — CHANGES_REQUESTED** [review]

File: src/customRules/trailingCommaInterfaceRule.ts (hunk 5 above)
Comment (ismail-syed) [reply src/customRules/trailingCommaInterfaceRule.ts:4]: For some reason, after updating I kept getting 
```
ERROR: src/customRules/trailingCommaInterfaceRule.ts[20, 41]: variable 'TrailingCommaInterfaceWalker' used before declaration
```

File: src/rules/functionality.ts (hunk 8 above)
Comment (ismail-syed) [reply src/rules/functionality.ts]: The rationale behind it is `Missing elements are probably an accidentally duplicated comma.`

File: src/customRules/trailingCommaInterfaceRule.ts (hunk 5 above)
Comment (GoodForOneFare) [reply src/customRules/trailingCommaInterfaceRule.ts:4]: Ah, okay.  They must use looser rules than us.

File: src/rules/functionality.ts (hunk 8 above)
Comment (GoodForOneFare) [reply src/rules/functionality.ts]: I understand the rule.  I'm asking why it's not being enforced.

File: CHANGELOG.md (hunk 6 above)
Comment (ismail-syed) [reply CHANGELOG.md]: Would it be worth the effort to manually have them listed here as well?

File: CHANGELOG.md (hunk 7 above)
Comment (ismail-syed) [reply CHANGELOG.md]: Fixed 006583d

File: CHANGELOG.md (hunk 6 above)
Comment (GoodForOneFare) [reply CHANGELOG.md]: I don't understand your question.  To be clear, I'm saying:
> This list makes it look like we've enabled 7 new rules, but we've really enabled 5

i.e., should the list provide more context like:

```
* `no-invalid-template-strings`
* `no-sparse-array` (disabled)
```

... or are we happy with consumers digging in to see what is/isn't enforced by default?

Comment (ismail-syed) [reply CHANGELOG.md]: Ahhhh, I assumed you'd like to have each rules' value listed. 

I like the `(disabled)` suggestion. 👍

File: src/rules/functionality.ts (hunk 8 above)
Comment (ismail-syed) [reply src/rules/functionality.ts]: Enabled 3dbe709

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,6 +2,21 @@
...
+### Added
+- Added new `tslint` rules from [v5.0.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v500) and [v5.1.0](https://github.com/palantir/tslint/blob/master/CHANGELOG.md#v510) releases
+    * `no-invalid-template-strings`
+    * `no-sparse-arrays`
+    * `no-unnecessary-callback-wrapper`
+    * `no-reference-import`
+    * `prefer-template`
+    * `return-undefined` (disabled)
+- Added new `tslint-react` rules from [v2.6.0](https://github.com/palantir/tslint-react/releases/tag/2.6.0) release [(#61)](https://github.com/Shopify/tslint-config-shopify/pull/61)
+    * `jsx-no-bind` (tslint-react)
```
Comment (GoodForOneFare) [inline CHANGELOG.md]: What does the `(tslint-react)` mean here?  That's already covered by the parent bullet point.

Comment (ismail-syed) [reply CHANGELOG.md]: Oh yes, forgot to remove it after refactoring this section.

Comment (ismail-syed) [reply CHANGELOG.md]: Fixed ce85081

**Review by GoodForOneFare — APPROVED** [review]
I was so happy when this PR just appeared out of thin air 👍 Thank you for acting like an owner!

**Comment by ismail-syed** [comment]
😄  ❤️

**Pull request merged**